### PR TITLE
Visit table finish

### PR DESCRIPTION
**Pull request by Johngtka**
Features checklist:
- [X] `added hiding of Re-Checked button when length of storing row id point is >= 1` 
- [X] `added swap of selected rows in visit table`
- [X] `user can deselect row after swap`
- [X] `added form field shows when re-check button clicked`
- [X] `when paginator has next page and user select tome point then on last page of table show Re-Checked button`
- [X] `when paginator has previous page then hide Re-Check`

ezgif com-video-to-gif

### CODE REVIEW

**Comment by Johngtka** [comment]
to fix:
- [] `displaying of text input when it is not needed`

**Comment by Johngtka** [comment]
S&M commit proposal:
`feat: swaping of selected points to one table on re-check button clicked`

**Comment by macrog** [comment]
> S&M commit proposal: `feat: swaping of selected points to one table on re-check button clicked`

_feat: swaping of selected points to one table on re-check button clicked_ -  this is to narrow, try to describe the feature you were working on more than what it was done, try again!

File: src/app/visit/visit.component.ts [hunk 1]
```diff
@@ -38,25 +38,23 @@ export class VisitComponent implements OnInit {
...
         'image',
     ];
-    dataSource:
-        | Store[]
-        | MatTableDataSource<Store, MatTableDataSourcePaginator>;
+    dataSource: MatTableDataSource<Store, MatTableDataSourcePaginator>;
     visitPoints: number[] = [];
+    store: Store[];
     selected = false;
+    checked = false;
```
Comment (macrog) [inline src/app/visit/visit.component.ts:45]: what is `checked` for what is used?

File: src/app/visit/visit.component.ts [hunk 2]
```diff
@@ -113,4 +120,12 @@ export class VisitComponent implements OnInit {
     ): object is Patient {
         return Object.hasOwn(object, 'name');
     }
+    checkVisit(): void {
```
Comment (macrog) [inline src/app/visit/visit.component.ts:123]: maybe better name for the function would be `createVisitPointsTable` ? or something like this? `checkVisit` doesn't describe much I would change the name.

**Comment by macrog** [comment]
When I select more than 10 rows in the first table the second table doesn't open on the first page?! 
Have a look at the gif, I have selected 23 rows and than clicked CHECK button, the next table should open on the 1st page and it opens on the last one

brave_CC1jVJWIsp

File: src/app/visit/visit.component.html [hunk 3]
```diff
@@ -71,13 +64,21 @@
...
+        <mat-label>{{ 'PATIENT_VISIT.NOTES' | translate }}</mat-label>
+        <input matInput />
+    </mat-form-field>
+    <button [hidden]="!checked" class="add-btn" mat-raised-button color="primary">
+        FINISH
+    </button>
+    <button [hidden]="!selected" (click)="checkVisit()" class="add-btn" mat-raised-button color="primary">
+        CHECK
+    </button>
+</div>
```
Comment (macrog) [inline src/app/visit/visit.component.html:84]: prettier the code!

File: src/app/visit/visit.component.html [hunk 4]
```diff
@@ -71,13 +64,21 @@
...
     <mat-paginator [pageSizeOptions]="[10]" aria-label="Select page">
     </mat-paginator>
 </div>
+<div class="finishBar" [hidden]="!patient.name">
+    <mat-form-field class="notesOnCheck" [ngClass]="{ notesOffCheck: !checked }">
+        <mat-label>{{ 'PATIENT_VISIT.NOTES' | translate }}</mat-label>
+        <input matInput />
+    </mat-form-field>
+    <button [hidden]="!checked" class="add-btn" mat-raised-button color="primary">
+        FINISH
```
Comment (macrog) [inline src/app/visit/visit.component.html:79]: This should be translated

File: src/app/visit/visit.component.html [hunk 5]
```diff
@@ -71,13 +64,21 @@
...
+    </button>
+    <button [hidden]="!selected" (click)="checkVisit()" class="add-btn" mat-raised-button color="primary">
+        CHECK
```
Comment (macrog) [inline src/app/visit/visit.component.html:82]: This should be translated

File: src/app/visit/visit.component.html [hunk 6]
```diff
@@ -71,13 +64,21 @@
...
     </table>
     <mat-paginator [pageSizeOptions]="[10]" aria-label="Select page">
     </mat-paginator>
 </div>
+<div class="finishBar" [hidden]="!patient.name">
+    <mat-form-field class="notesOnCheck" [ngClass]="{ notesOffCheck: !checked }">
+        <mat-label>{{ 'PATIENT_VISIT.NOTES' | translate }}</mat-label>
+        <input matInput />
+    </mat-form-field>
+    <button [hidden]="!checked" class="add-btn" mat-raised-button color="primary">
```
Comment (macrog) [inline src/app/visit/visit.component.html:78]: add a click method and print out `note` and `visitPoints` value in a console log , check **Example of the flow** in trello

**Review by macrog — CHANGES_REQUESTED** [review]

**Comment by macrog** [comment]
selecting a row on the last page of the table doesn't show the CHECK button 
image

**Comment by macrog** [comment]
this css almost has no effect it can be removed
image


| with css | without css |
| --- | --- |
| image  | image |

**Review by macrog — CHANGES_REQUESTED** [review]

File: src/app/visit/visit.component.ts (hunk 1 above)
Comment (Johngtka) [reply src/app/visit/visit.component.ts:45]: it used for hiding and showing check and finish button but i change this names

File: src/app/visit/visit.component.ts (hunk 2 above)
Comment (Johngtka) [reply src/app/visit/visit.component.ts:123]: done

File: src/app/visit/visit.component.html [hunk 7]
```diff
@@ -71,13 +64,21 @@
...
-            [ngClass]="{ selectedRow: visitPoints.indexOf(row.id) > -1 }"
-        ></tr>
+        <tr mat-row *matRowDef="let row; columns: displayedColumns" (click)="clickedRow(row)"
+            [ngClass]="{ selectedRow: visitPoints.indexOf(row.id) > -1 }"></tr>
     </table>
     <mat-paginator [pageSizeOptions]="[10]" aria-label="Select page">
     </mat-paginator>
 </div>
+<div class="finishBar" [hidden]="!patient.name">
+    <mat-form-field class="notesOnCheck" [ngClass]="{ notesOffCheck: !checked }">
```
[the remark this answers is not in the document]
Comment (Johngtka) [reply src/app/visit/visit.component.html]: when i use ngif or hidden it isn't work correctly and input still showing

[the remark this answers is not in the document]
Comment (macrog) [reply src/app/visit/visit.component.html]: this works: 
image

and the notesOffCheck class can be removed from css

File: src/app/visit/visit.component.html (hunk 6 above)
Comment (macrog) [reply src/app/visit/visit.component.html]: it should be printing `visitPoints`  not the full array of objects

**Comment by macrog** [comment]
you should add to a paginator a page event and a local method, like this:
image

and in a local method check exactly the same what are you checking for a logic to show CHECK button (this.rowSelect - btw I would chenge this name to showCheck the same as you have showFinish).  This way we make sure that even if user navigates to last page with a mouse click, not an arrow press, we will show the CHECK button.

File: src/app/visit/visit.component.css [hunk 8]
```diff
@@ -29,6 +29,10 @@
     height: 50px;
 }
 
+.notesOffCheck {
```
Comment (macrog) [inline src/app/visit/visit.component.css:32]: https://github.com/Johngtka/Biomagnethick-app/pull/35#discussion_r1184745460

File: src/app/visit/visit.component.html [hunk 9]
```diff
@@ -78,6 +78,37 @@
...
+        (page)="pageTriggerManually()"
+        [pageSizeOptions]="[10]"
+        aria-label="Select page"
+    >
     </mat-paginator>
 </div>
+<div [hidden]="!patient.name">
+    <mat-form-field
+        class="notesOnCheck"
+        [ngClass]="{ notesOffCheck: !showFinish }"
```
Comment (macrog) [inline src/app/visit/visit.component.html:91]: https://github.com/Johngtka/Biomagnethick-app/pull/35#discussion_r1184745460

File: src/app/visit/visit.component.ts [hunk 10]
```diff
@@ -73,23 +72,31 @@ export class VisitComponent implements OnInit {
         const index = this.visitPoints.indexOf(row.id);
         if (index !== -1) {
             this.visitPoints.splice(index, 1);
+            this.showCheck = false;
         } else {
             this.visitPoints.push(row.id);
         }
+        if (!this.paginator.hasNextPage() && this.visitPoints.length >= 1) {
+            this.showCheck = true;
+        }
```
Comment (macrog) [inline src/app/visit/visit.component.ts:81]: this block of code is repeated 3 times, it is a good moment to move it to a private function and call it 3 times the function

File: src/app/visit/visit.component.ts [hunk 11]
```diff
@@ -113,4 +126,24 @@ export class VisitComponent implements OnInit {
...
+
+    pageTriggerManually(): void {
+        if (!this.paginator.hasNextPage() && this.visitPoints.length >= 1) {
+            this.showCheck = true;
+        }
+    }
+
+    printNote(): void {
+        console.log(this.noteVal, this.dataSource.data);
+    }
```
Comment (macrog) [inline src/app/visit/visit.component.ts:148]: I would like to keep the private functions always at the bottom, first all public and then all private, please move the private function to the bottom of the file

**Review by macrog — CHANGES_REQUESTED** [review]
just a little bit more

File: src/app/visit/visit.component.html (hunk 7 above)
[the remark this answers is not in the document]
Comment (Johngtka) [reply src/app/visit/visit.component.html]: done

File: src/app/visit/visit.component.ts (hunk 10 above)
Comment (Johngtka) [reply src/app/visit/visit.component.ts:81]: done

**Review by macrog — CHANGES_REQUESTED** [review]
Something doesn't work well, steps to reproduce, select any number of rows, go to the last page, click CHECK button, in a new table deselect one of the rows, CHECK button is shown!

brave_Bthd8ePxe9

**Review by macrog — APPROVED** [review]
- [X] show CHECK button at the end of first selection
- [X] show FINISH button at the end of second selection
- [X] user can deselect any row in second table
- [X] note input added
- [X] FINISH button print outs values of notes and rows (missing to add visitPoints)

**Comment by Johngtka** [comment]
'''

**Pull request merged**